### DISCUSSION TITLE
SH-RPI Issue when PWM Activated

### DISCUSSION THREAD

**Discussion by DJSturgess** [hatlabs/discussions#108] [Q&A]
I need some help understanding why SH-RPI stops working correctly when PWM is enabled on the RPi.

My system comprises an RPi 3B+, a Waveshare 2-Channel Isolated CAN Hat and a SH-RPI.  For software it’s running Signalk server.  This all works perfectly until I enable PWM to drive a case fan.  The fan control also works perfectly, but it interrupts the correct operation of the SH-RPI.

With PWM active at 25kHz the regular 4 second flash of the SH-RPI LEDs becomes highly erratic, often flashing at much longer intervals than 4 seconds, and shutdowns often fail to happen correctly.  Correct operation resumes when the PWM signal is stopped.

I have tested this on new installations of both Bullseye and Bookworm.

The issue occurs with hardware-PWM using Python/pigpio, hardware-PWM using sysfs, and software-PWM using Python/pigpio.  It occurs even if the fan is not connected.   The issue does not occur if the PWM signal is created by bit-bashing the GPIO pin.

For hardware-PWM `dtoverlay=pwm,pin=12,func=4` is in config.txt.  The problems also occur with `dtoverlay=pwm-2chan,pin=12,func=4`.

This simple Python script enables hardware PWM using pigpio driving the fan at 50% on GPIO 12 at 25kHz and causing the SH-RPI issue.

```
gpio = 12             # Pin 32 
frequency = 25000     # 25kHz
dutycycle = 500000    # 50% of 1,000,000
import pigpio
pi=pigpio.pi()
pi.hardware_PWM(gpio, frequency, dutycycle)
```
This simple Bash script enables hardware PWM using sysfs driving the fan at 50% on GPIO 12 at 25kHz and causing the SH-RPI issue.

```
# 25kHz = 40,000 nS period and 20,000nS duty-cycle (50%)
echo 0 >     /sys/class/pwm/pwmchip0/export
echo 40000 > /sys/class/pwm/pwmchip0/pwm0/period
echo 20000 > /sys/class/pwm/pwmchip0/pwm0/duty_cycle
echo 1 >     /sys/class/pwm/pwmchip0/pwm0/enable
```

This simple Python script enables software PWM using pigpio driving the fan at 50% on GPIO 12 at 25kHz and causing the SH-RPI issue.

```
gpio = 12             # Pin 32 
frequency = 25000     # 25kHz
dutycycle = 50        # 50% of 100
import pigpio
pi=pigpio.pi()
pi.set_PWM_range(gpio,100)
pi.set_PWM_frequency(gpio, frequency)
pi.set_PWM_dutycycle(gpio, dutycycle)
```
The problem occurs when appropriate alternative pins are used. 

I would welcome any insight on why enabling PWM impacts the correct operation of SH-RPI and how to correct it.

**Comment by mairas** [comment]
That does sound really surprising. To me it'd seem like enabling PWM messes up with I2C communications. As to why that might be, I don't yet have an idea. You mentioned that it still happens when alternative pins are used - did you try to use PWM1?

FWIW, I just concocted simple fan control for the [HALPI](https://shop.hatlabs.fi/products/halpi-m-cm4-08-512-asc). My solution didn't use PWM, though - it does a simple bang-bang (on/off) control:

    # Enable fan control on the Waveshare CM4-IO-BASE-A board.
    dtoverlay=gpio-fan,gpiopin=18,temp=60000

Or actually, given that the `gpio-fan` overlay can be adjusted to do PWM control as well and it's using a PWM-capable pin, I don't know whether it already does that behind the scenes. The repo for that stuff is here: https://github.com/hatlabs/openplotter-pi-gen/blob/op4-halpi/stage7/00-enable-i2c/files/config.txt.part

**Reply by DJSturgess** [reply]
I’ve just tried swapping to PWM1 and all seems to work ok.  I’ll test it further to make sure it’s fully working.  Thanks for the suggestion.  I’d tried many scenarios, but hadn’t tried moving to PWM1.

I’d still like to understand why it doesn’t work with PWM0.

**Reply by mairas** [reply]
Me too. Quick Googling didn't provide any clear reasoning. I wonder if the I2C uses the same hardware clock as PWM0? Would be interesting to know.
